### DISCUSSION TITLE
Cloudnativepg failed to install on a namespace

### DISCUSSION THREAD

**Discussion by Kamal-Villupuram** [cloudnative-pg/cloudnative-pg#260] [General]
Hi,

I am trying to install cloudnativepg operator on a namespace and it failed with below errors. Access for API servers is not allowed in my organization. So how to include operator on the namespace and how to limit the cnpg-manager service account to access only the namespaces provided?

$ kubectl apply -f  cnpg-1.15.0.yaml
namespace/postgreppoc-development-business unchanged
serviceaccount/cnpg-manager unchanged
configmap/cnpg-default-monitoring unchanged
service/cnpg-webhook-service unchanged
deployment.apps/cnpg-controller-manager unchanged
Error from server (Forbidden): error when retrieving current configuration of:
Resource: "apiextensions.k8s.io/v1, Resource=customresourcedefinitions", GroupVersionKind: "apiextensions.k8s.io/v1, Kind=CustomResourceDefinition"
Name: "backups.postgresql.cnpg.io", Namespace: ""
from server for: "cnpg-1.15.0.yaml": customresourcedefinitions.apiextensions.k8s.io "backups.postgresql.cnpg.io" is forbidden: User "4facd75f-2b0a-49ae-8bdb-cabe9fa11104" cannot get resource "customresourcedefinitions" in API group "apiextensions.k8s.io" at the cluster scope: access denied
Error from server (Forbidden): error when retrieving current configuration of:
Resource: "apiextensions.k8s.io/v1, Resource=customresourcedefinitions", GroupVersionKind: "apiextensions.k8s.io/v1, Kind=CustomResourceDefinition"
Name: "clusters.postgresql.cnpg.io", Namespace: ""

**Comment by sxd** [comment]
Hi @Kamal-Villupuram 

Sadly, the YAML file provided on the release it's to be used to install the operator for the whole cluster meaning that it cannot be used to be installed in only one namespace to watch only that namespace, probably you can try with our Helm chart https://github.com/cloudnative-pg/charts but don't know if that capability is implemented, the best person to answer that is @phisco can you give us a hand here?

Regards!

**Comment by Kamal-Villupuram** [comment]
Thanks a lot for your quick reply. 
@phisco, can you please advise.

**Comment by phisco** [comment]
:wave: @Kamal-Villupuram, unfortunately, the helm chart is not at the moment able to handle installations on a single namespace, mainly because it's a difficult thing to do properly and some cluster-wide resources such as CustomResourceDefinitions can not be easily installed multiple times. 
Moreover, in your case, it looks like you don't have permission to read, nor create, CustomResourceDefinitions (CRDs), which are strongly required to be able to then create Cluster resources and, as I was saying above, are cluster-wide resources. I suggest you contact your cluster admin and see if it's possible to proceed with a cluster-wide installation if possible.

**Comment by Kamal-Villupuram** [comment]
Thanks. Can you please let me know the exact permissions required, so I can check with my cluster admin.  
below is the reply from my admin. 

Cluster admin permissions cannot be granted to other teams apart from platform operations teams in any cluster (non-prod and prod) so we need to work together with the vendor to narrow down the permissions and provide permissions to the CRD's that they need to manage and not capabilities to modify any cluster resources that are part of the cluster.

**Reply by gbartolini** [reply]
Hi @Kamal-Villupuram,

The operator automates in a declarative way a lot of operations related to PostgreSQL management that otherwise would require manual and imperative interventions. Such operations also include security related matters at RBAC (e.g. service accounts), pod (e.g. security context constraints) and Postgres levels (e.g. TLS certificates). It should be seen as an extension of Kubernetes. Please look at here for details: https://cloudnative-pg.io/documentation/1.15.1/security/#role-based-access-control-rbac

Another useful source is the documentation we have prepared at EDB for the OpenShift operator, which explains how we believe permissions should be set. Let us know if that applies to your case: https://www.enterprisedb.com/docs/postgres_for_kubernetes/latest/openshift/#users-and-permissions

**Comment by phisco** [comment]
Pasting here the answer I gave you on slack, @Kamal-Villupuram, let us know if you have other doubts:

Given how RBAC works in Kubernetes, to install the operator, you’ll need at least the same permissions the operator will have, see the ClusterRole definition in the official [release manifest](https://github.com/cloudnative-pg/cloudnative-pg/blob/main/releases/cnpg-1.15.1.yaml), moreover you will also need to be able to apply any resource defined in that manifest, which means mainly CustomResourceDefinitions. For these reasons, operator’s installation usually require admin permissions.
